### DISCUSSION TITLE
Configure importModuleSpecifier for tsserver

### DISCUSSION THREAD

**Discussion by hoi4** [LazyVim/LazyVim#1124] [Q&A] [answered]
Hi @folke 

Thank you for this amazing starting point to nvim! ❤️ 

I am new to nvim and struggling with configuring `tsserver` to set the `importModuleSpecifier` setting. I am using a NX monorepo and need to change the setting from `"shortest"` to `"project-relative"`. 

Using `:Mason` revealed the following setting, but it didn't work for me, so I guess I am using it wrong

```
→ typescript.preferences.importModuleSpecifier                                   default: "shortest"
```

Any help would be greatly appreciated 🙏🏻

**Comment by dpetka2001** [comment]
Could you try in a file in `/lua/plugins/` directory the following
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        tsserver = {
          settings = {
            typescript = {
              preferences = {
                importModuleSpecifier = "project-relative",
              },
            },
          },
        },
      },
    },
  },
}
```

**Reply by hoi4** [reply]
Thank you for your suggestion. This was also the configuration that I tried when I found the `:Mason` docs about `tsserver`. Unfortunately, this doesn't seem to work for me 🤔 

It still uses the value `"shortest"` for the setting.

**Reply by jyuan0** [reply]
Maybe it's `importModuleSpecifierPreference`?

**Reply by hoi4** [reply]
Unfortunately not

**Reply by hoi4** [reply]
I have tried to use [typescript-tools.nvim](https://github.com/pmizio/typescript-tools.nvim) just to see whether I can configure the `moduleImportSpecifier` option there and it works. Unfortunately that causes new issues with e.g. eslint not working anymore when saving a buffer.

At this point it seems I am just not able to make it work 🙄

**Comment by folke** [comment, marked as answer]
You need to set 
```lua
lspconfig.opts.servers.tsserver.init_options = {
    hostInfo = "neovim",
    preferences = {
      includeCompletionsForModuleExports = true,
      includeCompletionsForImportStatements = true,
      importModuleSpecifierPreference = "relative",
    },
  },
```

**Reply by hoi4** [reply]
Thank you! 👍🏻

**Reply by mbackermann** [reply]
Where do I put this?

**Reply by mbackermann** [reply]
I tried adding this to the plugins folder, but I can't get it to work. Whenever I import something, it always uses the absolute path. Can you help me here @hoi4 ?

**Reply by rogeriofrsouza** [reply]
You can paste that code on your `neovim-lspconfig` lua file.
It may be something like: `lua/plugins/lspconfig.lua`

```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        tsserver = {
          init_options = {
            preferences = {
              importModuleSpecifierPreference = "relative",
              -- other settings
            },
          },
        },
      },
    },
  },
}
```

**Reply by covelloz** [reply]
Absolutely wonderful. Nice thread 👏! This config almost should almost be added to the [nvim-lspconfig section](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#tsserver)! It's very helpful for those of us who have path aliases defined in tsconfig & want to have the tsserver LSP code action use non-relative paths (i.e use the path aliases). Odd that is it not a configurable setting in the `setup` function 🤔

**Reply by asd-ez** [reply]
For me works (`vtsls`)
```lua
-- /lua/plugins/lsp.lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        vtsls = {
          settings = {
            typescript = {
              preferences = {
                includeCompletionsForModuleExports = true,
                includeCompletionsForImportStatements = true,
                importModuleSpecifier = "non-relative",
              },
            },
          },
        },
      },
    },
  },
}
```

**Reply by beetol** [reply]
Do you have to have typescript-language-server installed along with vtsls?

**Reply by timredd** [reply]
This worked perfectly. Thanks!

**Reply by pierodev0** [reply]
Thanks

**Reply by hr0dv1tn1r** [reply]
I'm using Nvim with nvchad and currently not making this working. I tried in lua/custom/configs/lspconfig.lua and lua/custom/plugins.lua. Neither i'ts working. Can someone help me?

<img width="880" height="603" alt="image" src="https://github.com/user-attachments/assets/252c6ef7-6a03-41e3-be40-9f0c7e8a5318" />

**Comment by grantralls** [comment]
Slightly off topic but I'm curious, I've had a very difficult time diving into the tsserver source to discover what options are available. How did you use mason to discover them?

**Reply by hoi4** [reply]
use `:Mason` > find the LSP client > press Enter > `LSP server configuration schema (press enter to collapse)` > press Enter

I found it [here](https://www.reddit.com/r/neovim/comments/120r1ye/comment/jdpa8k8/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)
